privacy-pools/libs/lean-imt/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+#![no_std]
+
+use soroban_sdk::{
+    symbol_short, vec, BytesN, Env, Symbol, Vec, U256, Map,
+    crypto::bls12_381::Fr as BlsScalar,
+};
+use poseidon::Poseidon255;
+
+/// Storage keys for the LeanIMT
+pub const TREE_ROOT_KEY: Symbol = symbol_short!("root");
+pub const TREE_DEPTH_KEY: Symbol = symbol_short!("depth");
+pub const TREE_LEAVES_KEY: Symbol = symbol_short!("leaves");
+
+/// Converts u64 to BlsScalar for test compatibility
+pub fn u64_to_bls_scalar(env: &Env, value: u64) -> BlsScalar {
+    BlsScalar::from_u256(U256::from_u32(env, value as u32))
+}
+
+/// Converts BlsScalar to BytesN<32> for Soroban storage
+pub fn bls_scalar_to_bytes(scalar: BlsScalar) -> BytesN<32> {
+    scalar.to_bytes()
+}
+
+/// Converts BytesN<32> to BlsScalar for computation
+pub fn bytes_to_bls_scalar(bytes_n: &BytesN<32>) -> BlsScalar {
+    BlsScalar::from_bytes(bytes_n.clone())
+}
+
+/// Lean Incremental Merkle Tree implementation with hybrid approach:
+/// - Internal computation uses BlsScalar for perfect Circom compatibility
+/// - Storage and API uses BytesN<32> for Soroban compatibility
+pub struct LeanIMT<'a> {
+    env: &'a Env,
```

Comment:
You don't need to store env reference here, you can just clone it (it is just an `Rc`).